File: src/main/Java/boj_step16/boj4949.java
import java.io.*;
import java.util.Stack;

public class boj4949 {
    public static void main(String[] args) throws IOException {
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(System.out));
        String line;
        while (!(line = reader.readLine()).equals(".")) writer.append((isVps(line)) ? "yes\n" : "no\n");
        writer.close();
    }
    private static boolean isVps(String line) {
        Stack<Boolean> list = new Stack<>();
        for (char c: line.toCharArray())
            switch (c) {
                case ')': if (list.isEmpty() || !list.pop()) return false; break;
                case ']': if (list.isEmpty() || list.pop()) return false; break;
                case '(': list.push(true); break;
                case '[': list.push(false); break;
            }
        if (list.isEmpty()) return true;
        else return false;
    }
}

File: src/main/kotlin/boj_step16/boj4949.kt
import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.util.*

fun main() {
    val r = BufferedReader(InputStreamReader(System.`in`))
    val w = BufferedWriter(OutputStreamWriter(System.out))
    while (true) {
        val s = r.readLine()
        if (s.equals(".")) break
        w.appendLine(if (isVPS(s)) "yes" else "no")
    }
    w.close()
}

private fun isVPS(s: String): Boolean {
    val list = Stack<Boolean>()
    s.forEach {
        when (it) {
            ')' -> if (list.isEmpty() || !list.pop()) return false
            ']' -> if (list.isEmpty() || list.pop()) return false
            '(' -> list.push(true)
            '[' -> list.push(false)
        }
    }
    if (list.isEmpty()) return true
    else return false
}